src/serializer/functions.js

```diff
@@ -264,34 +267,26 @@ export class Functions {
       effects = additionalFunctionEffects.effects;
       this.writeEffects.set(functionValue, additionalFunctionEffects);
 
-      // look for newly registered optimized functions
-      let modifiedProperties = additionalFunctionEffects.effects.modifiedProperties;
       // Conceptually this will ensure that the nested additional function is defined
       // although for later cases, we'll apply the effects of the parents only.
       this.realm.withEffectsAppliedInGlobalEnv(() => {
-        for (let [propertyBinding] of modifiedProperties) {
-          let descriptor = propertyBinding.descriptor;
-          if (descriptor && propertyBinding.object === optimizedFunctionsObject) {
-            let newValue = descriptor.value;
-            invariant(newValue instanceof Value); //todo: this does not seem invariantly true
-            let newEntry = this.__optimizedFunctionEntryOfValue(newValue);
-            if (newEntry) {
-              additionalFunctions.add(newEntry.value);
-              getEffectsFromAdditionalFunctionAndNestedFunctions(newEntry.value, newEntry.argModelString);
-              // Now we have to rember the stack of effects that need to be applied to deal with
-              // this additional function.
-            }
-          }
+        let newOptFuncs = this.__generateInitialOptimizedFunctionsFromRealm();
+        for (let newEntry of newOptFuncs) {
+          additionalFunctions.add(newEntry.value);
+          getEffectsFromAdditionalFunctionAndNestedFunctions(newEntry.value, newEntry.argModel);
         }
+        // Now we have to rember the stack of effects that need to be applied to deal with
+        // this additional function.
         return null;
       }, additionalFunctionEffects.effects);
+      this.realm.optimizedFunctions.unshift(...oldRealmOptimizedFunctions);
       invariant(additionalFunctionStack.pop() === functionValue);
       for (let t2 of this.realm.tracers) t2.endOptimizingFunction(currentOptimizedFunctionId);
     };
 
     while (additionalFunctionsToProcess.length > 0) {
```

Comment:
This while loop with shifting... Took me a while to realize that this is now a simple look with no queuing going on, since that's handled by the recursive nature of `getEffectsFromAdditionalFunctionAndNestedFunctions`? Simplify.